apis/Google.Cloud.Firestore.V1/Google.Cloud.Firestore.V1/FirestoreClientPartial.cs

```diff
@@ -0,0 +1,124 @@
+            ApplyResourcePrefixHeader(ref settings, request.Parent);
+
+        partial void Modify_RollbackRequest(ref RollbackRequest request, ref CallSettings settings) =>
+            ApplyResourcePrefixHeader(ref settings, request.Database);
+
+        partial void Modify_RunQueryRequest(ref RunQueryRequest request, ref CallSettings settings) =>
+            ApplyResourcePrefixHeader(ref settings, request.Parent);
+
+        partial void Modify_UpdateDocumentRequest(ref UpdateDocumentRequest request, ref CallSettings settings) =>
+            ApplyResourcePrefixHeader(ref settings, request.Document?.Name);
+
+        private static void ApplyResourcePrefixHeader(ref CallSettings settings, string resource)
+        {
+            // If we haven't been given a resource name, just let the request as it is.
+            if (string.IsNullOrEmpty(resource))
+            {
+                return;
+            }
+            string database = GetDatabaseResourceName(resource);
+            settings = settings.WithHeader(ResourcePrefixHeader, database);
+        }
+        
+        // Visible for testing
+
+        /// <summary>
+        /// Retrieves the database resource name from a full resource name.
+        /// Validation is performed as far as the database ID but no further; the database ID is deemed to end at the first slash.
+        /// </summary>
+        /// <param name="resource">The resource name, which must start with projects/{project_id}/databases/{database_id}</param>
+        /// <returns>The database resource name</returns>
+        internal static string GetDatabaseResourceName(string resource)
+        {
+            const string projectsPrefix = "projects/";
+            const string databasesPrefix = "databases/";
+            if (string.CompareOrdinal(resource, 0, projectsPrefix, 0, projectsPrefix.Length) != 0)
+            {
+                // "projects/" doesn't match
+                ThrowInvalidResource();
+            }
+            int endOfProjectId = resource.IndexOf('/', projectsPrefix.Length);
+            if (endOfProjectId == -1 || endOfProjectId == projectsPrefix.Length)
+            {
+                // Empty project ID or no slash at the end of it
+                ThrowInvalidResource();
+            }
+            if (string.CompareOrdinal(resource, endOfProjectId + 1, databasesPrefix, 0, databasesPrefix.Length) != 0)
+            {
+                // "databases/" doesn't match
+                ThrowInvalidResource();
+            }
+            int startOfDatabaseId = endOfProjectId + 1 + databasesPrefix.Length;
+            if (startOfDatabaseId == resource.Length)
+            {
+                // No database ID
+                ThrowInvalidResource();
+            }
+            int endOfDatabaseId = resource.IndexOf('/', startOfDatabaseId);
+            // It's valid for the whole resource name to be the database name
```

Comment:
nit: I find this comment confusing here. Should it be on top of the return statement instead?

Reply:
Moved.